File: python-app/routes/pdf_routes.py
from fastapi import APIRouter, HTTPException
from fastapi.responses import JSONResponse
from pdf2image import convert_from_path
import os
import time
from pydantic import BaseModel
# Using the alternative text extractor that doesn't rely on PyMuPDF
# from utils.text_extractor_alt import extract_text_from_pdf
from utils.text_extractor import get_text_from_pdf, convert_pdf_to_txt

router = APIRouter()

class PdfRequest(BaseModel):
    pdf_path: str

@router.post("/convert-pdf")
def convert_pdf(request: PdfRequest):
    """
    Endpoint to convert PDF to image
    """
    try:
        pdf_path = request.pdf_path
        
        # Handle relative paths correctly inside the container
        # If the path doesn't start with /, assume it's relative to /app
        if not pdf_path.startswith('/'):
            actual_path = os.path.join('/app', pdf_path)
        else:
            actual_path = pdf_path
        
        print(f"Looking for PDF at: {actual_path}")
        
        # Check if the file exists
        if not os.path.exists(actual_path):
            raise HTTPException(status_code=404, detail=f"PDF file not found: {pdf_path} (resolved to {actual_path})")
        
        # Create output directory if it doesn't exist
        output_dir = os.path.join(os.path.dirname(actual_path), "converted_images")
        os.makedirs(output_dir, exist_ok=True)
        
        # Generate a unique filename using timestamp
        timestamp = int(time.time())
        output_image_path = os.path.join(output_dir, f"converted_{timestamp}.jpg")
        
        # Convert PDF to image (first page only)
        images = convert_from_path(actual_path, first_page=1, last_page=1)
        
        # Save the image
        if images:
            images[0].save(output_image_path, 'JPEG')
            return {"image_path": output_image_path}
        else:
            raise HTTPException(status_code=500, detail="Failed to convert PDF to image")
    
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error processing PDF: {str(e)}")


@router.post("/extract-text")
def extract_text(request: PdfRequest):
    """
    Endpoint to extract text and patient details from PDF
    
    This endpoint extracts both structured patient data and raw text from a PDF file.
    """
    try:
        pdf_path = request.pdf_path
        
        # Handle relative paths correctly inside the container
        if not pdf_path.startswith('/'):
            actual_path = os.path.join('/app', pdf_path)
        else:
            actual_path = pdf_path
        
        print(f"Looking for PDF at: {actual_path}")
        
        # Check if the file exists
        if not os.path.exists(actual_path):
            raise HTTPException(status_code=404, detail=f"PDF file not found: {pdf_path} (resolved to {actual_path})")
        
        # First convert PDF to text
        text_content = convert_pdf_to_txt(actual_path)
        
        # Then extract structured data from the text content
        result = get_text_from_pdf(text_content)
        
        # Parse the result to get the JSON data
        # The result is in format "[(status)]{json_data}"
        import json
        import re
        
        # Extract status and JSON data
        match = re.match(r'\[\((\d)\)\](.*)', result)
        if match:
            status = match.group(1)
            json_str = match.group(2)
            
            # Parse the JSON string into a Python dict
            parsed_data = json.loads(json_str)
            
            # Check which critical fields are missing
            missing_fields = []
            critical_fields = ["Patient", "Patient ID", "Date of Birth", "Test Type"]
            for field in critical_fields:
                if not parsed_data.get(field):
                    missing_fields.append(field)
            
            # Format the response as requested
            if status == "1":
                message = "PDF text extraction successful"
            else:
                if missing_fields:
                    message = f"PDF text extraction partially successful. Missing critical fields: {', '.join(missing_fields)}"
                else:
                    message = "PDF text extraction partially successful, some data may be missing"
            
            return {
                "data": {
                    "text": parsed_data
                },
                "message": message
            }
        else:
            return {
                "data": {
                    "text": {}
                },
                "message": "Error parsing PDF extraction result"
            }
    
    except Exception as e:
        return {
            "data": {
                "text": {}
            },
            "message": f"Error extracting text from PDF: {str(e)}"
        } 

File: node-app/routes/pdf-routes.js
const express = require('express');
const axios = require('axios');
const path = require('path');

const router = express.Router();

/**
 * Endpoint to convert PDF to image
 * Expects a JSON body with { "pdfPath": "path/to/pdf" }
 */
router.post('/convert-pdf', async (req, res) => {
  try {
    const { pdfPath } = req.body;
    
    if (!pdfPath) {
      return res.status(400).json({ error: 'PDF path is required' });
    }
    
    // Make a request to the Python FastAPI service
    const pythonResponse = await axios.post('http://python-app:5000/api/convert-pdf', {
      pdf_path: pdfPath
    });
    
    // Return the image path from the Python service
    return res.json({ 
      success: true, 
      imagePath: pythonResponse.data.image_path,
      message: 'PDF successfully converted to image'
    });
  } catch (error) {
    console.error('Error converting PDF:', error.message);
    
    // If the error comes from the Python service, forward its response
    if (error.response && error.response.data) {
      return res.status(error.response.status || 500).json({
        success: false,
        error: error.response.data.detail || 'Error processing PDF'
      });
    }
    
    return res.status(500).json({ 
      success: false, 
      error: 'Failed to convert PDF to image'
    });
  }
});

/**
 * Endpoint to extract text from PDF
 * Expects a JSON body with { "pdfPath": "path/to/pdf" }
 * Returns both structured data and raw text extracted from the PDF
 */
router.post('/extract-text', async (req, res) => {
  try {
    const { pdfPath } = req.body;
    
    if (!pdfPath) {
      return res.status(400).json({ error: 'PDF path is required' });
    }
    
    // Make a request to the Python FastAPI service
    const pythonResponse = await axios.post('http://python-app:5000/api/extract-text', {
      pdf_path: pdfPath
    });
    
    const responseData = pythonResponse.data;
    
    // Check if the Python service encountered an error
    if (responseData.error) {
      return res.status(422).json({
        success: false,
        error: responseData.error,
        message: 'Error extracting text from PDF'
      });
    }
    
    // Return the complete data from the Python service, including raw_text
    return res.json({ 
      success: responseData.status, 
      data: responseData.data,
      raw_text: responseData.raw_text,
      message: responseData.status 
        ? 'PDF text successfully extracted' 
        : 'PDF text extraction partially successful, some data may be missing'
    });
  } catch (error) {
    console.error('Error extracting text from PDF:', error.message);
    
    // If the error comes from the Python service, forward its response
    if (error.response && error.response.data) {
      return res.status(error.response.status || 500).json({
        success: false,
        error: error.response.data.detail || 'Error extracting text from PDF'
      });
    }
    
    return res.status(500).json({ 
      success: false, 
      error: 'Failed to extract text from PDF'
    });
  }
});

/**
 * Endpoint to extract visualization images from PDF
 * Expects a JSON body with { "pdfPath": "path/to/pdf" }
 * Returns URLs to access the extracted images
 */
router.post('/extract-images', async (req, res) => {
  try {
    const { pdfPath } = req.body;
    
    if (!pdfPath) {
      return res.status(400).json({ error: 'PDF path is required' });
    }
    
    // Make a request to the Python FastAPI service
    const pythonResponse = await axios.post('http://python-app:5000/api/extract-images', {
      pdf_path: pdfPath
    });
    
    const responseData = pythonResponse.data;
    
    // Check if the Python service encountered an error
    if (responseData.error) {
      return res.status(422).json({
        success: false,
        error: responseData.error,
        message: 'Error extracting images from PDF'
      });
    }
    
    // Transform the Python static URLs to be accessible through the Node.js app
    if (responseData.data && responseData.data.images) {
      responseData.data.images = responseData.data.images.map(image => {
        // Replace /static/ with /python-static/ to match our proxy setup
        const nodeUrl = image.url.replace('/static/', '/python-static/');
        return {
          ...image,
          url: nodeUrl
        };
      });
    }
    
    // Return the image data from the Python service with updated URLs
    return res.json({
      success: true,
      data: responseData.data,
      message: responseData.message
    });
  } catch (error) {
    console.error('Error extracting images from PDF:', error.message);
    
    // If the error comes from the Python service, forward its response
    if (error.response && error.response.data) {
      return res.status(error.response.status || 500).json({
        success: false,
        error: error.response.data.detail || 'Error extracting images from PDF'
      });
    }
    
    return res.status(500).json({ 
      success: false, 
      error: 'Failed to extract images from PDF'
    });
  }
});

module.exports = router; 